### DISCUSSION TITLE
Warp Engines

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#230] [Feature Proposals]
_For more information about the ship systems discussions, see https://github.com/Thorium-Sim/thorium-nova/discussions/224_

# Warp Engines

Warp engines are use to travel long distances, such as between planets and solar systems. Using space magic, they propel the ship faster than the speed of light. Warp speed is so unbelievably fast that the ship effectively disappears from short range sensors while traveling at that speed. This means that battles or ship-to-ship interactions only happens at sub-light speeds.

## Warp Factors
Since warp speeds are so absurdly fast, we categorize them by warp factors. A warp engine could have any number of warp factors, but it's most common to have 5 warp factors. For brevity, we'll assume 5 warp factors for this document.

Warp engines also have "Emergency" and "Destructive" warp factors which extrapolate the top warp factor speed. These speeds are discouraged due to high power usage and high heat generation.

Warp 5 is considered cruising speed, which is the fastest speed the ship can safely travel. Warp 1 is the slowest warp speed, and is determined by multiplying the cruising speed by some ratio, typically 0.01. If Warp 5 is 29980000 km/s (100c) then Warp 1 is 299800 km/s, or the speed of light.

Warp factors between 1 and 5 are extrapolated using a linear function. If Warp 1 were 1km/s and Warp 5 were 5km/s, then warp 3 would be 3km/s.

## Solar System vs Interstellar Travel

While 100c is fast enough to get from Earth to Mars in 10 seconds, traveling to the nearest solar system (Alpha Centauri, 4ly away) would take 16 days, far too long for a casual player. Because of this, warp travel is much slower within a solar system than in interstellar space. If cruising speed in a solar system is 100c, then cruising speed in interstellar space could be something like 2000000c. At that speed, a ship could travel to Alpha Centauri in 70 seconds. That may seem too fast, but remember, we may need to travel to many solar systems over the course of a game.

Also, the high-warp speeds only work outside of a solar system, so the ship must first travel at the slower speeds until they leave the solar system (the boundary of the solar system is just past the furthest planet in that solar system), then at high interstellar speeds, and then at solar system speeds again until they reach the planet in their destination solar system.

So in total, a trip from Earth to an assumed planet in Alpha Centauri would take 6 minutes total - a reasonable, perhaps even long, travel time. 

## The Warp Dynamo and Starting the Engines

It might be tempting to use warp engines frequently to make quick getaways, but for storytelling purposes, it's better that there is a delay between when you intend to activate the warp engines and when the ship speeds away.

To provide this delay, we say that warp engines require incredibly high current to activate - higher than the reactors can all provide simultaneously. Because of this, activating the warp engines first requires power be stored in a [Compensated Pulsed Dynamo](https://en.wikipedia.org/wiki/Compensated_pulsed_alternator) (We replace "Alternator" with "Dynamo" in the name, because "Warp Dynamo" sounds cooler). The Warp Dynamo is a separate system altogether consisting of a massive flywheel that is slowly spun up to high speeds, and then rapidly braked to immediately convert that stored energy into high current. The process of charging the warp dynamo is called "spooling".

Starting the Warp Engines begins with making sure the Warp Dynamo is connected to the warp engines via the power distribution screen. The Dynamo is then spooled to the level necessary to jump to Warp 1. Then the "Activate Dynamo" button is pressed, the dynamo discharges, and the ship jumps to warp.

> It should also be possible to connect the Warp Dynamo to other systems, like the phasers or shields, where pressing the dynamo button overpower those systems, causing them to be super effective for a moment before damaging them from the high current.

Increasing warp factor requires the dynamo to be spooled and activated again for each warp factor. The time to spool to Warp 1 should take around 30 seconds, and each subsequent spooling should take 5 seconds or so, depending on the power available to the dynamo. Playtesting can determine whether these times are good, or whether jumping to each warp separately is a good idea.

To add to the drama, jumps to Warp 1 only happen while the ship is at full stop - impulse engines must be deactivated.

## Power & Heat

Warp engines require relatively little power when activated. Higher warp factors require higher heat, with Emergency and Destructive warp requiring exponentially higher power.

Spooling the Warp Dynamo takes all excess power that is currently routed to the dynamo. That means activating the engines requires a lot of power over a long period of time, depending on how fast the crew wants to spool the dynamo. The high power consumption of the Warp Dynamo is to discourage continuously spooling in hopes of making a quick getaway. 

Spooling the dynamo continuously generates heat, to further discourage continuous spooling. Activating the Dynamo also produces a large amount of heat suddenly. To avoid damaging the Warp Dynamo, it needs to be sufficiently cooled before and after activating.

**Comment by Froginator45** [comment]
In case people haven't seen this graphic before (it may have some debatable incorrect info but it's still a good idea for how the science fiction realm defined different speeds):
![image](https://user-images.githubusercontent.com/19561600/155055533-4c6df0d2-559a-4839-a202-88ea5c149d9d.png)

**Comment by Ryan-Anderson-PhD** [comment]
You know... there is another option.

Impulse could be for maneuvering near planets and between planets and moons. Warp could be for travel between planets. You could use a hyperspace/jump drive of sorts for interstellar travel. 

There is a lot of real science this could be based off of. From oversimplified version of it from string theory is that by travelling through some other dimensions, you could essentially create a short cut between linear point a and point b in three dimensions. 

But, hyperspace would be difficult, because unless you had a warp field active while you were in it, everything about you and the ship would also expand into those extra dimensions, which is something you would not survive. 

So, hyperspace travel could be much faster, but could only occur outside of systems, and would require a warp field to be active.

Just a thought.

**Reply by 1deejay** [reply]
Messing with the Nova system and navigating I really like this idea. Hyper jumps could also be done as long as the way is clear. You of course can't jump through a planet.

The good part is that it makes battles at impulse speed actually viable speed wise. Since torpedoes going at full impulse speed as it is now would be catastrophically destructive to planets. And they would have to go faster than impulse to catch up a fleeing ship.

**Comment by Froginator45** [comment]
I'm also wondering about the jump drive (or wormhole) being fast travel points within the universe. Immobile stations that are where they are because of the enormous energy requirements for making wormhole travel possible. I think it could introduce more storyline elements and simplify the gameplay a bit without having so many different types of engines. You could still have a jump drive of sorts that lets your ship be able to make the wormhole journey without being destroyed. Without it, you would have major problems due to the inherent instability of wormholes. Just another thought. :)

**Comment by alexanderson1993** [comment]
I'm going to finalize this discussion. We'll implement it as described in the initial post. Jump drives are a possible form of locomotion that we can implement at a later time, including limitations around using jump drive inside solar systems. We could also create fast-travel locations in space. Both of those can be discussed in separate discussions.
